### ISSUE TITLE
Dilation Factors

### ISSUE THREAD

**Issue by denabazazian** [AbdulMoqeet/HRAN#4]
Thanks for sharing your code. I have some questions about the dilation factors. 
In your code, you just considered the dilation factors as 1 and 2. 
I am just wondering if you have tried some other factors? and would it be possible to change the code in order to have various dilation factors?
For instance, if I just add a new dilation factor as 3, then I should modify [hran.py](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py) cod between the lines [#27](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#27) and [#53](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#53) as below:

        self.conv_3 = conv(n_feats, n_feats, kernel_size_1)
        self.conv_3_2 = conv(n_feats, n_feats, kernel_size_1, dilation=2)
        self.conv_3_3 = conv(n_feats, n_feats, kernel_size_1, dilation=3)

        self.conv_3_1 = conv(n_feats*2, n_feats, kernel_size_1)
        self.conv_3_2_1 = conv(n_feats*2, n_feats, kernel_size_1, dilation=2)
        self.conv_3_3_1 = conv(n_feats*2, n_feats, kernel_size_1, dilation=3)

        self.LR = nn.LeakyReLU(inplace=True)
        self.conv_11 = conv(n_feats*2, n_feats, 1)

    def forward(self, x):
        res_x = x

        a =  self.conv_du_1(x)
        b1 = self.LR(self.conv_3(x))
        b2 = self.LR(self.conv_3_2(x)) + b1
        b3 = self.LR(self.conv_3_3(x)) + b2
        B = torch.cat([ b1, b2, b3],1)

        b1 = self.conv_3_1(B)
        b2 = self.LR(self.conv_3_2_1(B)) + b1
        b3 = self.LR(self.conv_3_3_1(B)) + b2

        B = torch.cat([b1,b2,b3], 1)
        B = self.conv_11(B)

My main questions are followings: 
1. Is there a way to make the list of dilation factors in a dynamic manner instead of having the hard code?
2. Why did you use the `nn.AdaptiveAvgPool2d(1)` in [self.conv_du_1](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#L19)?
2. Why do you define the reduction in line [#17](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#17) as 4?
3. What is the reason of `self.conv_3_1` and `self.conv_3_2` in lines [#30](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#39) and [#31](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#31)? and why in these lines the `n_feats` multiplied by 2? Also, according to `self.conv_3_1` and `self.conv_3_2`, what is the reason of having the second `b1` and `b2` in lines [#48](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#48) and [#49](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/hran.py#49)?
4. When you define the convolution in [common.py](https://github.com/AbdulMoqeet/HRAN/blob/master/HRAN/src/model/common.py#L7), you defined the padding as `(kernel_size//2)*dilation`, why did you divide the kernel size to two?

**Comment by AbdulMoqeet** [comment]
Hello! 

First of all, thank you for your interest in our work. 

I did try other dilation factors as well but they didn't work for me, however, I think, if I would have added dilation factors with another layer like you did, it would have worked. But, I wanted to make it simple, so, I just used two layers (with dilation factors 1 and 2) but you may use others as well. 

Below are the replies to your questions:

**1.  Is there a way to make the list of dilation factors in a dynamic manner instead of having the hard code?**
If you mean, which dilation factor should be used for the model, then currently, I only think about NAS (Neural Architecture Search) which may help to decide about the dilation factor. 

**2. Why did you use the nn.AdaptiveAvgPool2d(1) in self.conv_du_1?**
It is used for global average pooling where ```(1)``` is the desired output i.e., ```(batch_size,num_channels,1,1)```

**3.  Why do you define the reduction in line #17 as 4?** 
It comes from empirical results. The previous methods have shown that it should be 16 but, I found out that for image restoration, we lose information with a high reduction factor. So, it might be feasible for image classification tasks, but not for image restoration tasks. 

Thus, instead of using ```reduction=16```, we chose the ```reduction=num_features/16``` in our work. 

**4. What is the reason of self.conv_3_1 and self.conv_3_2 in lines #30 and #31? and why in these lines the n_feats multiplied by 2? Also, according to self.conv_3_1 and self.conv_3_2, what is the reason of having the second b1 and b2 in lines #48 and #49?**

I found out that previous layers were not enough. So, I added more layers (namely, self.conv_3_1 and self.conv_3_2). Whereas, additional layers are represented by ```b1``` and ```b2``` which are just naming conventions, used for adding another block. 

**5. When you define the convolution in common.py, you defined the padding as ```(kernel_size//2)*dilation```, why did you divide the kernel size to two?**
It is required to get the desired padding to have the same size output.

**Comment by denabazazian** [comment]
Thanks a lot @AbdulMoqeet for your quick and thorough reply, it was very useful to me.
